core/internal/runfiles/uploader.go

```diff
@@ -119,3 +156,105 @@ func (u *uploader) lockForOperation(method string) bool {
 
 	return true
 }
+
+// Uploads the given files unless we are offline.
+//
+// This increments `uploadWG` and returns immediately. The wait group is
+// signalled as file uploads finish.
+func (u *uploader) upload(relativePaths []string) {
+	if u.settings.IsOffline() {
+		return
+	}
+
+	relativePaths = u.filterNonExistingAndWarn(relativePaths)
+	u.uploadWG.Add(len(relativePaths))
+
+	go func() {
+		createRunFilesResponse, err := gql.CreateRunFiles(
+			context.Background(),
+			u.graphQL,
+			u.settings.GetEntity(),
+			u.settings.GetProject(),
+			u.settings.GetRunID(),
+			relativePaths,
+		)
+		if err != nil {
+			u.logger.CaptureError("runfiles: CreateRunFiles returned error", err)
+			u.uploadWG.Add(-len(relativePaths))
+			return
+		}
+
+		if len(createRunFilesResponse.CreateRunFiles.Files) != len(relativePaths) {
+			u.logger.CaptureError(
+				"runfiles: CreateRunFiles returned unexpected number of files",
+				nil,
+				"expected",
+				len(relativePaths),
+				"actual",
+				len(createRunFilesResponse.CreateRunFiles.Files),
+			)
+			u.uploadWG.Add(-len(relativePaths))
+			return
+		}
+
+		for _, f := range createRunFilesResponse.CreateRunFiles.Files {
+			if f.UploadUrl == nil {
+				u.logger.CaptureWarn(
+					"runfiles: CreateRunFiles has empty UploadUrl",
+					"response",
+					createRunFilesResponse,
+				)
+				u.uploadWG.Done()
+				continue
+			}
+
+			u.scheduleUploadTask(
+				f.Name,
+				*f.UploadUrl,
+				createRunFilesResponse.CreateRunFiles.UploadHeaders,
+			)
+		}
+	}()
+}
+
+// Warns for any non-existing files and returns a slice without them.
+func (u *uploader) filterNonExistingAndWarn(relativePaths []string) []string {
+	existingRelativePaths := make([]string, 0)
+
+	for _, relativePath := range relativePaths {
+		localPath := filepath.Join(u.settings.GetFilesDir(), relativePath)
+
+		if _, err := os.Stat(localPath); os.IsNotExist(err) {
+			u.logger.Warn("runfiles: upload: file does not exist", "path", localPath)
+		} else {
+			existingRelativePaths = append(existingRelativePaths, relativePath)
+		}
+	}
+
+	return existingRelativePaths
+}
+
+// Schedules a file upload task.
+//
+// Decrements `uploadWG` when the task is complete or fails.
+func (u *uploader) scheduleUploadTask(
+	relativePath string,
+	uploadURL string,
+	headers []string,
+) {
+	localPath := filepath.Join(u.settings.GetFilesDir(), relativePath)
+	task := &filetransfer.Task{
+		Type:    filetransfer.UploadTask,
+		Path:    localPath,
+		Name:    relativePath,
+		Url:     uploadURL,
+		Headers: headers,
+	}
+
+	// TODO: FileTransferInfoRequest on progress and completion
+	task.SetCompletionCallback(func(t *filetransfer.Task) {
+		u.uploadWG.Done()
+	})
```

Comment:
btw, this filetransferinfo was hacked to get the progress and completion, feel free to modify it as long as we get the correct count and progress :)

Reply:
awesome, i'll take a look at how it's used and see if there's a neater way!